src/backend/distributed/executor/multi_router_executor.c

```diff
@@ -344,8 +425,20 @@ ExecuteTaskAndStoreResults(QueryDesc *queryDesc, Task *task,
 		 */
 		if (!gotResults && expectResults)
 		{
-			queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
-									   &currentAffectedTupleCount);
+			/* use try/catch to maintain txn participant list during fail-fast errors */
+			PG_TRY();
+			{
+				queryOK = StoreQueryResult(routerState, connection, tupleDescriptor,
+										   &currentAffectedTupleCount);
+			}
+			PG_CATCH();
+			{
+				/* this placement is already aborted; remove it from participant list */
+				participantList = list_delete_ptr(participantList, taskPlacement);
```

Comment:
Yeah, I checked out the implementation before using the list things I'm going here (in particular, was worried about quadratic performance when using deletion in loops), but in typical use the replication factor is probably two or three (does anyone use four?), so the optimization seemed premature.

I'll check out your suggestions.